### ISSUE TITLE
[RFC] efficient nan-checker for distributed-collectives

### ISSUE THREAD

**Issue by wconstab** [pytorch/pytorch#124348]
PyTorch already has a NAN checker ([autograd anomaly detector](https://pytorch.org/docs/stable/_modules/torch/autograd/anomaly_mode.html#detect_anomaly)) which is great if you need to identify which operator caused a NAN - usually to help you find a buggy part of your model code, but is also slow because it performs a cpu sync/copy to extract the NAN-ness after running every operator in your program.

For distributed programs, it can also be the case that bad hardware causes NANs and finding the source of the NANs is as simple as raising an exception or crashing the particular rank that first observed a NAN.  Since NANs can only flow between ranks/nodes by virtue of communication operations, it is possible to check for NANs only at the inputs to communications.  Further, the check may not even need to be synchronous, as long as we can raise an error sometime soon after observing the NAN, allowing it to be enabled more frequently (or even on-by-default).

Proposal:

1) find choke-points in C10D such as `ProcessGroupNCCL::collective(` and `pointToPoint` (maybe others too?) which let us add instrumentation in few places and cover all ops
2) add a helper which executes the equivalent of `found_nan = isnan(input)._is_any_true()`

From here, the proposal gets a bit hacky, ideally we could do it a more clean way.

3) register a new cuda event after recording `_is_any_true()` operation.  The event represents that the nan-check has asynchronously completed on gpu.  Bundle the event into a 'Work' object with an operator name "isnan"
 - somehow stuff the 'found_nan' tensor value into the Work too? Or stick it into some separate map
4) add a special path inside watchdogHandlerLoop where any work object which completes and has "isnan" operator name will be able to call logic like `if (found_nan.item()) { raise Exception("NAN found on pg x rank y") }`

Perhaps all we need to do to make this cleaner is to subclass Work such that the special subclass can do 
  - container for the 'found_nan' tensor
  - check_exception() method can be overridden to check the found_nan value and create the above mentioned exception 

Issues
- need to ensure we raise the error first on the rank where it happened.  Can't really gaurantee this if we do the check in an async thread, unless we take steps to ensure that we execute the check _before_ we start the following collective.  In other words, we can let the previous pytorch (compute) operators complete asynchronously to produce the 'input' tensor, but at that point we have to sync before we issue the collective.

cc @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @H-Huang @kwen2501 @awgu @penguinwu @fegin @XilunWu @wanchaol @fduwjj @wz337 @tianyu-l @yf225 @chauhang @d4l3k

**Comment by xw285cornell** [comment]
Just curious - if there are NaNs in the model, in the end it'll be surfaced as loss==NaN. Checking NaN at collective won't necessarily make it easier to detect which operator is the NaN first happens. So the purpose is to check which rank NaN happens first?

**Comment by wconstab** [comment]
right- the point is to find it before it spreads to other ranks. once it shows up in the loss in one rank, chances are it might show up in the loss in every rank, making it impossible to find the bad node.

**Comment by kwen2501** [comment]
> For distributed programs, it can also be the case that bad hardware causes NANs

What's the chance of that?

> it is possible to check for NANs only at the inputs to communications.

If that's the case, what would be the advantage of checking for NaN _inside_ a collective (whether at PyTorch level or NCCL level) than checking for NaN on an input before passing it to a collective?

Lastly, I think it is legal to perform collectives or P2P on NaNs, in particular, _Quiet_ NaNs, which are purposed for propagation. For example, there is nothing wrong in sending a Quiet NaN to another process. Even for arithmetic, as long as ieee has defined an operation for NaN, such reduction should be deemed performable. Some NaNs' purpose is exactly to propagate through all arithmetic operations, or in the collective sense, propagate through all ranks, without raising an exception.

**Comment by kwen2501** [comment]
I agree that user should have freedom, and such freedom can be in a variety of ways:

1. User may choose where to detect the NaN
- Before a collective (as proposed in this RFC); or
- After the collective, to see if any peer propagated a NaN to "me".

2. User may choose how to deal with the NaN, for example:
- Reaction to signals: some NaNs are used for signaling, so user may have their own logic when seeing such signal.
- Masking: some users may choose to mask out the NaNs (e.g. received from collectives) and continue with the computation.
- Detection: in the case of all-gather, user may use that to detect which rank is in a specific state.

All the the above is based on the program's semantics. 

Accordingly, I think the best place for such freedom is outside of collective calls, because it would be too hard for collectives to define APIs or environment variables to cover them.

**Comment by wconstab** [comment]
closed as completed by https://github.com/pytorch/pytorch/pull/125726

**Comment by houminz** [comment]
Hi, as #125726 shows we can enable NaN checker when we set ENV `TORCH_NCCL_NAN_CHECK`, I am wondering if there are any performance test comparing whether we enable NAN_CHECK at large scale training
cc  @wconstab @shuqiangzhang
